# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/html/HTMLSourceElement.h
# Commit: acdd013202e1
# Full Hash: acdd013202e1e688986e7e6c747b7c7fee38474e
# Author: Boris Chiou <boris.chiou@gmail.com>
# Date: 2022-08-23 02:46:57
# Regressor Bug: 1694741
# File Overlap Count: 1
# Description:
#   Bug 1694741 - Part 6: Map width/height attributes to the style of img elements. r=emilio
#   
#   This patch will use the width/height attributes from <source> to override
#   width/height/aspect-ratio CSS property values of <img> elements.
#   
# ==============================================================================

diff -r 2b4d9516b882 -r acdd013202e1 dom/html/HTMLSourceElement.h
--- a/dom/html/HTMLSourceElement.h	Mon Aug 22 20:18:39 2022 +0000
+++ b/dom/html/HTMLSourceElement.h	Mon Aug 22 20:18:39 2022 +0000
@@ -29,11 +29,13 @@
 
   NS_IMPL_FROMNODE_HTML_WITH_TAG(HTMLSourceElement, source)
 
-  virtual nsresult Clone(dom::NodeInfo*, nsINode** aResult) const override;
+  nsresult Clone(dom::NodeInfo*, nsINode** aResult) const override;
 
   // Override BindToTree() so that we can trigger a load when we add a
   // child source element.
-  virtual nsresult BindToTree(BindContext&, nsINode& aParent) override;
+  nsresult BindToTree(BindContext&, nsINode& aParent) override;
+
+  void UnbindFromTree(bool aNullParent) override;
 
   // If this element's media attr matches for its owner document.  Returns true
   // if no media attr was set.
@@ -97,6 +99,14 @@
     SetUnsignedIntAttr(nsGkAtoms::height, aHeight, 0, aRv);
   }
 
+  const nsMappedAttributes* GetAttributesMappedForImage() const {
+    return mMappedAttributesForImage;
+  }
+
+  static bool IsAttributeMappedToImages(const nsAtom* aAttribute) {
+    return aAttribute == nsGkAtoms::width || aAttribute == nsGkAtoms::height;
+  }
+
  protected:
   virtual ~HTMLSourceElement();
 
@@ -114,6 +124,11 @@
                         bool aNotify) override;
 
  private:
+  // Generates a new MediaList using the given input
+  void UpdateMediaList(const nsAttrValue* aValue);
+
+  void BuildMappedAttributesForImage();
+
   bool IsInPicture() const {
     return GetParentElement() &&
            GetParentElement()->IsHTMLElement(nsGkAtoms::picture);
@@ -128,8 +143,9 @@
   // The triggering principal for the srcset attribute.
   nsCOMPtr<nsIPrincipal> mSrcsetTriggeringPrincipal;
 
-  // Generates a new MediaList using the given input
-  void UpdateMediaList(const nsAttrValue* aValue);
+  // The mapped attributes to HTMLImageElement if we are associated with a
+  // <picture> with a valid <img>.
+  RefPtr<nsMappedAttributes> mMappedAttributesForImage;
 };
 
 }  // namespace mozilla::dom